File: shared/src/commonMain/kotlin/com/slowthemdown/shared/model/MeasurementSystem.kt
package com.slowthemdown.shared.model

enum class MeasurementSystem(val rawValue: String) {
    IMPERIAL("imperial"),
    METRIC("metric");

    companion object {
        fun fromRawValue(raw: String): MeasurementSystem =
            entries.firstOrNull { it.rawValue == raw } ?: IMPERIAL
    }
}

object UnitConverter {
    // Constants
    private const val MPS_TO_MPH = 2.23694
    private const val MPS_TO_KMH = 3.6
    private const val METERS_TO_FEET = 3.28084

    // Speed: m/s → display unit
    fun displaySpeed(mps: Double, system: MeasurementSystem): Double = when (system) {
        MeasurementSystem.IMPERIAL -> mps * MPS_TO_MPH
        MeasurementSystem.METRIC -> mps * MPS_TO_KMH
    }

    // Display unit → m/s
    fun speedToMps(value: Double, system: MeasurementSystem): Double = when (system) {
        MeasurementSystem.IMPERIAL -> value / MPS_TO_MPH
        MeasurementSystem.METRIC -> value / MPS_TO_KMH
    }

    fun speedUnit(system: MeasurementSystem): String = when (system) {
        MeasurementSystem.IMPERIAL -> "MPH"
        MeasurementSystem.METRIC -> "km/h"
    }

    // Distance: meters → display unit
    fun displayDistance(meters: Double, system: MeasurementSystem): Double = when (system) {
        MeasurementSystem.IMPERIAL -> meters * METERS_TO_FEET
        MeasurementSystem.METRIC -> meters
    }

    // Display unit → meters
    fun distanceToMeters(value: Double, system: MeasurementSystem): Double = when (system) {
        MeasurementSystem.IMPERIAL -> value / METERS_TO_FEET
        MeasurementSystem.METRIC -> value
    }

    fun distanceUnit(system: MeasurementSystem): String = when (system) {
        MeasurementSystem.IMPERIAL -> "ft"
        MeasurementSystem.METRIC -> "m"
    }

    // Calibration ratio: pixels per meter → display
    fun displayPixelsPerUnit(ppm: Double, system: MeasurementSystem): Double = when (system) {
        MeasurementSystem.IMPERIAL -> ppm / METERS_TO_FEET // px/m → px/ft
        MeasurementSystem.METRIC -> ppm
    }

    fun calibrationUnit(system: MeasurementSystem): String = when (system) {
        MeasurementSystem.IMPERIAL -> "px/ft"
        MeasurementSystem.METRIC -> "px/m"
    }
}

File: ios/SlowThemDown/Models/MeasurementSystem.swift
import Foundation

enum MeasurementSystem: String, CaseIterable, Codable {
    case imperial
    case metric

    static var deviceDefault: MeasurementSystem {
        let locale = Locale.current
        // US, Liberia, Myanmar use imperial
        let imperialRegions: Set<String> = ["US", "LR", "MM"]
        if let region = locale.region?.identifier, imperialRegions.contains(region) {
            return .imperial
        }
        return .metric
    }
}

enum UnitConverter {
    private static let mpsToMph = 2.23694
    private static let mpsToKmh = 3.6
    private static let metersToFeet = 3.28084

    static func displaySpeed(_ mps: Double, system: MeasurementSystem) -> Double {
        switch system {
        case .imperial: mps * mpsToMph
        case .metric: mps * mpsToKmh
        }
    }

    static func speedToMps(_ value: Double, system: MeasurementSystem) -> Double {
        switch system {
        case .imperial: value / mpsToMph
        case .metric: value / mpsToKmh
        }
    }

    static func speedUnit(_ system: MeasurementSystem) -> String {
        switch system {
        case .imperial: "MPH"
        case .metric: "km/h"
        }
    }

    static func displayDistance(_ meters: Double, system: MeasurementSystem) -> Double {
        switch system {
        case .imperial: meters * metersToFeet
        case .metric: meters
        }
    }

    static func distanceToMeters(_ value: Double, system: MeasurementSystem) -> Double {
        switch system {
        case .imperial: value / metersToFeet
        case .metric: value
        }
    }

    static func distanceUnit(_ system: MeasurementSystem) -> String {
        switch system {
        case .imperial: "ft"
        case .metric: "m"
        }
    }

    static func displayPixelsPerUnit(_ ppm: Double, system: MeasurementSystem) -> Double {
        switch system {
        case .imperial: ppm / metersToFeet
        case .metric: ppm
        }
    }

    static func calibrationUnit(_ system: MeasurementSystem) -> String {
        switch system {
        case .imperial: "px/ft"
        case .metric: "px/m"
        }
    }
}
